opnode/l1/downloader.go

```diff
@@ -2,291 +2,69 @@ package l1
-//
-// Receipt-downloading workers can be added/removed through AddReceiptWorkers at any time.
-type downloader struct {
-	// cache of ongoing/completed block tasks: block hash -> block
-	cacheEvict *lru.Cache
-	// Get/Add calls need to be atomic to avoid duplicate jobs being created and added
-	cacheLock sync.Mutex
-
-	receiptTasks       chan *receiptTask
-	receiptWorkers     []ethereum.Subscription
-	receiptWorkersLock sync.Mutex
-
-	// source to pull data from. Nil when downloader is closed
-	src DownloadSource
-}
-
-func NewDownloader(dlSource DownloadSource) Downloader {
-	dl := &downloader{
-		receiptTasks: make(chan *receiptTask, receiptQueueSize),
-		src:          dlSource,
-	}
-	evict := func(k, v interface{}) {
-		// stop downloading things if they were evicted (already finished items are unaffected)
-		v.(*downloadTask).Finish(wrappedErr{DownloadEvictedErr})
-		v.(*downloadTask).cancel()
-	}
-	dl.cacheEvict, _ = lru.NewWithEvict(cacheSize, evict)
-	return dl
-}
-
-func (dl *downloader) Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error) {
-	// check if we are already working on it
-	dl.cacheLock.Lock()
-	if dl.src == nil {
-		dl.cacheLock.Unlock()
-		return nil, nil, DownloadClosedErr
+func (dl Downloader) Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error) {
+	block, err := dl.client.BlockByHash(ctx, id.Hash)
+	if err != nil {
+		return nil, nil, err
 	}
-
-	var dlTask *downloadTask
-	if dlTaskIfc, ok := dl.cacheEvict.Get(id.Hash); ok {
-		dlTask = dlTaskIfc.(*downloadTask)
-	} else {
-		ctx, cancel := context.WithCancel(ctx)
-		dlTask = &downloadTask{ctx: ctx, cancel: cancel}
-		dl.cacheEvict.Add(id.Hash, dlTask)
-
-		// pull the block in the background
+	txs := block.Transactions()
+	receipts := make([]*types.Receipt, len(txs))
+
+	semaphoreChan := make(chan struct{}, MaxConcurrentFetchesPerCall)
+	defer close(semaphoreChan)
+	var retErr error
+	var errMu sync.Mutex
+	var wg sync.WaitGroup
+	for idx, tx := range txs {
+		wg.Add(1)
+		i := idx
+		hash := tx.Hash()
 		go func() {
-			ctx, cancel := context.WithTimeout(ctx, fetchBlockTimeout)
-			defer cancel()
-			bl, err := dl.src.BlockByHash(ctx, id.Hash)
-			if err != nil {
-				dlTask.Finish(wrappedErr{fmt.Errorf("failed to download block %s: %v", id.Hash, err)})
-				return
-			}
-
-			txs := bl.Transactions()
-			dlTask.block = bl
-			dlTask.receipts = make([]*types.Receipt, len(txs))
-
-			for i, tx := range txs {
-				dl.receiptTasks <- &receiptTask{blockHash: id.Hash, txHash: tx.Hash(), txIndex: uint64(i), dest: dlTask}
-			}
-
-			// no receipts to fetch? Then we are done!
-			if len(txs) == 0 {
-				dlTask.Finish(wrappedErr{nil})
+			semaphoreChan <- struct{}{}
+			for j := 0; j < MaxReceiptRetry; j++ {
+				receipt, err := dl.client.TransactionReceipt(ctx, hash)
+				if err != nil && j == MaxReceiptRetry-1 {
+					// dl.log.Error("Got error in final retry of fetch", "err", err)
+					errMu.Lock()
```

Comment:
Nit: Could use `atomic.Value` rather than a mutex, but it doesn't really make that much of a differentce.